### DISCUSSION TITLE
为什么这么设计系列文章 - 面向信仰编程 · /whys-the-design

### DISCUSSION THREAD

**Discussion by draveness** [draveness/blog-comments#609] [Announcements]
https://draveness.me/whys-the-design/

**Comment by yezihack** [comment]
**为什么这么设计**
为什么这么设计系列? 让我们思考更多,不只是看到技术的表面,而更深入的了解技术的背后.值得学习这种思维模式.

**Comment by he-sb** [comment]
【001 为什么 TCP 建立连接需要三次握手】这篇文章指向的链接错误，指向了 `https://draveness.me/whys-the-design-redis-single-thread/` 这篇文章

**Comment by draveness** [comment]
> 【001 为什么 TCP 建立连接需要三次握手】这篇文章指向的链接错误，指向了 `https://draveness.me/whys-the-design-redis-single-thread/` 这篇文章

好的，我来改一下

**Comment by iapolog2u** [comment]
期待后面会不会有 f1/spanner等分布式关系型数据库相关的博文 ：）

**Comment by draveness** [comment]
> 期待后面会不会有 f1/spanner等分布式关系型数据库相关的博文 ：）

之前写过其他分布式数据库的相关的文章，但是 F1 和 Spanner 没写过

**Comment by androidZzT** [comment]
作者所有的文章思路都非常清晰。能否之后分享一篇关于，你遇到一个问题是之后是如何查询资料，汲取资料中的精华，然后总结出结论，最后呈现出一篇漂亮的文章的

**Comment by draveness** [comment]
> 作者所有的文章思路都非常清晰。能否之后分享一篇关于，你遇到一个问题是之后是如何查询资料，汲取资料中的精华，然后总结出结论，最后呈现出一篇漂亮的文章的

最近好多人提了这个要求，不过得过一段时间

**Comment by Scyano** [comment]
我就想知道你用德莱文上到多少分了

**Comment by AviorAlong** [comment]
为什么这么设计 缓存和缓冲来一篇呗

**Comment by draveness** [comment]
> 为什么这么设计 缓存和缓冲来一篇呗

可以的，不过说不定会放到另一个系列里

**Comment by mayli** [comment]
不错，下次我给别人解释一个东西的时候就直接发这个链接就好了:)

**Comment by discipline-yyl** [comment]
为什么我问不出那么多为什么？
为什么我回答不出那么多为什么？
努力，向大神看齐，没有捷径，干就对了

**Comment by WenHaiGo** [comment]
您好  从文章中学了好多，也推荐给了朋友，我也想写写博客， 如何搭建一个这样的网站呢

**Comment by gaoryrt** [comment]
为什么 DNS 是一个应用层协议而不是数据链路层的呢？

**Comment by tinycedar** [comment]
大佬你这个博客是用的哪个模版？

**Comment by yixiaotian1230** [comment]
> @tinycedar
> 大佬你这个博客是用的哪个模版？


帮大佬答一下：hugo。皮肤自选https://www.gohugo.org/theme/

**Comment by vc60er** [comment]
弱弱的请教一个为什么，
为什么20行能返回1，21行返回0。难道map是指针类型，slice是一个结构类型。

```go
package main

import "fmt"

func test_map(m map[int]int) {
	m[0] = 1
}

func test_slice(a []int) {
	a = append(a, 1)
}

func main() {
	m := make(map[int]int)
	a := make([]int, 0, 10)

	test_map(m)
	test_slice(a)

	fmt.Println(len(m)) // 1
	fmt.Println(len(a)) // 0
}

```

**Comment by liuyu121** [comment]
> @vc60er
> 弱弱的请教一个为什么，
> 为什么20行能返回1，21行返回0。难道map是指针类型，slice是一个结构类型。
> 
> ```go
> package main
> 
> import "fmt"
> 
> func test_map(m map[int]int) {
> 	m[0] = 1
> }
> 
> func test_slice(a []int) {
> 	a = append(a, 1)
> }
> 
> func main() {
> 	m := make(map[int]int)
> 	a := make([]int, 0, 10)
> 
> 	test_map(m)
> 	test_slice(a)
> 
> 	fmt.Println(len(m)) // 1
> 	fmt.Println(len(a)) // 0
> }
> 
> ```

搞清 `map` 与 `slice` 的区别，就一目了然了。

`map` 本质是个指针 `*hmap`，但 `slice` 却是个值拷贝，所以 `test_slice` 虽然修改了底层数组，但是并没有修改原来的 `slice。`

**Comment by daisuki-konbini** [comment]
什么时候可以开个 kubernetes 相关的新坑那就太好了，希望作者可以详细介绍下kubernetes 网络原理。

**Comment by Queeee** [comment]
作者真是突然发现的宝藏博主哈哈哈，希望什么时候博主可以介绍一篇关于学习方法的文章~🤣

**Comment by lukelmouse-github** [comment]
这个博客是怎么做的，真的好看啊！！！

**Comment by GnailiyLin** [comment]
【011 为什么 MongoDB 使用 B 树】这篇文章打不开了

**Comment by PegasusWang** [comment]
赞，希望可以把文章导出成电子书，零散时间阅读。给作者打赏!

**Comment by AlanMaxxx** [comment]
我喜欢这些“为什么”！！

**Comment by chapsticks** [comment]
为什么博主会这么多

**Comment by blacksheepaul** [comment]
"Why the"是我做梦都想看的主题系列，请加大更新力度

**Comment by JASONPANGGO** [comment]
非常喜欢～ 从技术方案原始的设计思路角度来解释技术的原理，理解世界运行的方式

**Comment by Smityz** [comment]
> @draveness
> > 作者所有的文章思路都非常清晰。能否之后分享一篇关于，你遇到一个问题是之后是如何查询资料，汲取资料中的精华，然后总结出结论，最后呈现出一篇漂亮的文章的
> 
> 最近好多人提了这个要求，不过得过一段时间

同求！大佬太强了

**Comment by umelly** [comment]
文章「为什么 MongoDB 使用 B 树」循环跳转，无法访问，链接地址
https://draveness.me/whys-the-design-mongodb-b-tree/

**Comment by draveness** [comment]
> 文章「为什么 MongoDB 使用 B 树」循环跳转，无法访问，链接地址
> https://draveness.me/whys-the-design-mongodb-b-tree/

这篇文章下掉了

**Comment by lanrenbulan** [comment]
博主可否说说mysql的 null。网上搜索的结果基本都是不推荐使用，但为什么null又存在，而且列又默认为null

**Comment by draveness** [comment]
> @lanrenbulan
> 博主可否说说mysql的 null。网上搜索的结果基本都是不推荐使用，但为什么null又存在，而且列又默认为null

这个问题挺好的

**Comment by RickHe-biubiubiu** [comment]
为什么德莱文转行做程序员了？

**Comment by tzongw** [comment]
https://draveness.me/whys-the-design-mongodb-b-tree/404.html

为什么这篇文章打不开？提示ERR_TOO_MANY_REDIRECTS

**Comment by gua-pian** [comment]
佩服作者，难得的优秀博文

**Comment by Jigsawk** [comment]
赞一个

**Comment by peanutzhen** [comment]
为什么epoll用红黑树？

**Comment by EddieChan1993** [comment]
作者你好，我很想听你讲一下IO多路复用这些相关的知识，这块知识网上虽然很多，但是我就想听你来一个系列

**Comment by vc60er** [comment]
催更一下
